### PR TITLE
arm64: dts: qcom: msm8916-samsung-e7: Add initial dts

### PR DESCRIPTION
**Pull request by wonderfulShrineMaidenOfParadise**
This dts adds support for Samsung Galaxy E7 smartphone released in 2015.

Unfortunately, some variants were released with outdated 32-bit only
firmware and never received any update from Samsung. Since the 32-bit
TrustZone firmware is signed there seems to be no way currently to
actually boot this device tree on arm64 Linux at the moment. :(

However, it is possible to use this device tree by compiling an ARM32 kernel
instead. The device tree can be easily built on ARM32 with an #include
and it works really well there. To avoid confusion for others it is still
better to add this device tree on arm64. Otherwise it's easy to forget
to update this one when making some changes that affect all MSM8916 devices.

Maybe someone finds a way to boot ARM64 Linux on this device at some point.
In this case I expect that this device tree can be simply used as-is.

E7 looks like a chimera of A5 and S4 Mini VE. Keeping a note here if something can be shared.
A5 traits:
* `&sdhc_2` is removable
* `&tkey_en_default` at GPIO 97
* `s3fwrn5` NFC
* `a2015` audio codec

S4 Mini VE traits:
* `mpss@86800000` and `gps@8c200000`
* `sm5504` MUIC
* `wcn3620`

Todo:
- [ ] msm8916-mainline/linux-panel-drivers#9
- [ ] msm8916-mainline/lk2nd#144
- [ ] Upstream 083b312

### CODE REVIEW

**Comment by wonderfulShrineMaidenOfParadise** [comment]
https://github.com/msm8916-mainline/linux-panel-drivers/pull/9 is created to automatically generate the panel driver.
`ea8061-id400418` is the used one.

File: drivers/gpu/drm/panel/msm8916-generated/Makefile [hunk 1]
```diff
@@ -11,6 +11,7 @@ obj-$(CONFIG_DRM_PANEL_MSM8916_GENERATED) += panel-motorola-harpia-tianma.o
 obj-$(CONFIG_DRM_PANEL_MSM8916_GENERATED) += panel-motorola-osprey-inx.o
 obj-$(CONFIG_DRM_PANEL_MSM8916_GENERATED) += panel-motorola-surnia-boe.o
 obj-$(CONFIG_DRM_PANEL_MSM8916_GENERATED) += panel-oppo-15009-nt35592-jdi.o
+obj-$(CONFIG_DRM_PANEL_MSM8916_GENERATED) += panel-samsung-ea8061-id400418.o
```
Comment (wonderfulShrineMaidenOfParadise) [inline drivers/gpu/drm/panel/msm8916-generated/Makefile:14]: https://github.com/msm8916-mainline/linux-panel-drivers/pull/9 is created to automatically generate the panel driver.

File: drivers/input/touchscreen/stmfts.c [hunk 2]
```diff
@@ -325,9 +325,11 @@ static int stmfts_command(struct stmfts_data *sdata, const u8 cmd)
 	if (err)
 		return err;
 
-	if (!wait_for_completion_timeout(&sdata->cmd_done,
-					 msecs_to_jiffies(1000)))
-		return -ETIMEDOUT;
```
Comment (wonderfulShrineMaidenOfParadise) [inline drivers/input/touchscreen/stmfts.c:330]: It looks not right to drop it completely. Maybe raise the timeout interval.

Comment (wonderfulShrineMaidenOfParadise) [reply drivers/input/touchscreen/stmfts.c:330]: It seems there is non-true stuff returned with `&sdata->cmd_done`. Need some help from the author.
I don't see `mms114` or `zinitix` using this, though.

File: drivers/input/touchscreen/stmfts.c [hunk 3]
```diff
@@ -254,6 +255,7 @@ static void stmfts_parse_events(struct stmfts_data *sdata)
 		case STMFTS_EV_CONTROLLER_READY:
 		case STMFTS_EV_SLEEP_OUT_CONTROLLER_READY:
 		case STMFTS_EV_STATUS:
+		case STMFTS_EV_HACK:
```
Comment (wonderfulShrineMaidenOfParadise) [inline drivers/input/touchscreen/stmfts.c:258]: `0x80` is received during init but not handled. Consider it expected.

File: drivers/input/touchscreen/stmfts.c [hunk 4]
```diff
@@ -325,11 +327,12 @@ static int stmfts_command(struct stmfts_data *sdata, const u8 cmd)
...
-					 msecs_to_jiffies(1000)))
-		return -ETIMEDOUT;
-
-	return 0;
+	int i;
+	for (i=0; i < 2; i++)
+		if (wait_for_completion_timeout(&sdata->cmd_done,
+			msecs_to_jiffies(1)) == 0)
+			return 0;
+	return -ETIMEDOUT;
```
Comment (wonderfulShrineMaidenOfParadise) [inline drivers/input/touchscreen/stmfts.c:335]: It seems that it's the timeout length, but the count of attempts `wait_for_completion_timeout()`.
Some commands require at least 2 attempts to complete.
https://paste.debian.net/1227992/

File: drivers/input/touchscreen/stmfts.c [hunk 5]
```diff
@@ -553,9 +553,13 @@ static int stmfts_power_on(struct stmfts_data *sdata)
...
 	msleep(50);
 
+	/* HACK There is no event received during STMFTS_SYSTEM_RESET.
+	 * Ignoring the timeout.
+	 */
 	err = stmfts_command(sdata, STMFTS_SYSTEM_RESET);
 	if (err)
-		return err;
+		dev_warn(&sdata->client->dev,
+			 "failed to perform system reset: %d\n", err);
```
Comment (wonderfulShrineMaidenOfParadise) [inline drivers/input/touchscreen/stmfts.c:562]:
```
postmarketos-msm8916:~$ dmesg |grep stmfts
[    0.512347] stmfts 6-0049: starting command 0x80
[    0.537347] stmfts 6-0049: event 0x80 received
[    0.537360] stmfts 6-0049: unknown event 0x80
[    0.537366] stmfts 6-0049: event 0x0 received
[    0.572385] stmfts 6-0049: starting command 0xa0
[  102.960609] stmfts 6-0049: failed to perform system reset: -110
[  102.961314] stmfts 6-0049: starting command 0x91
[  102.986226] stmfts 6-0049: event 0x11 received
[  102.987093] stmfts 6-0049: Received event in 1000 ms
[  102.987714] stmfts 6-0049: starting command 0xa3
[  103.401054] stmfts 6-0049: event 0x16 received
[  103.401807] stmfts 6-0049: Received event in 1000 ms
[  103.402433] stmfts 6-0049: starting command 0xa4
[  103.490250] stmfts 6-0049: event 0x16 received
[  103.491004] stmfts 6-0049: Received event in 1000 ms
[  103.491626] stmfts 6-0049: starting command 0xa2
[  103.541450] stmfts 6-0049: event 0x16 received
[  103.542203] stmfts 6-0049: Received event in 1000 ms
[  103.543840] input: stmfts as /devices/platform/soc/78b9000.i2c/i2c-6/6-0049/input/input4
```
Waiting for a reply from the author of stmfts, Andi, to understand how to receive an event.

**Comment by wonderfulShrineMaidenOfParadise** [comment]
A potential solution https://github.com/msm8916-mainline/linux/pull/215/commits/38627bd622cc7cda4317f2381f02b45f479a41bb has been found.
Reference: https://lxr.trvn.ru/linux_qcom_msm8916/SM-E7000/source/drivers/input/touchscreen/stm_fts/fts_ts.c#L283

File: arch/arm64/boot/dts/qcom/msm8916-samsung-e7.dts [hunk 6]
```diff
@@ -102,6 +126,28 @@
 	};
 };
 
+&dsi0 {
+	panel@0 {
+		compatible = "samsung,e7-panel";
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-samsung-e7.dts:178]: https://github.com/msm8916-mainline/lk2nd/pull/144 is created for panel selection.

**Review by wonderfulShrineMaidenOfParadise — COMMENTED** [review]
Upstreaming for stmfts has been planned.

**Review by stephan-gh — CHANGES_REQUESTED** [review]

**Review by stephan-gh — APPROVED** [review]

**Review by TravMurav — APPROVED** [review]

**Comment by stephan-gh** [comment]
Integrated into `rebase42` branch, thanks!

**Pull request closed without merging**